### ISSUE TITLE
No Service Worker support

### ISSUE THREAD

**Issue by fregante** [fczbkk/css-selector-generator#162]
I'm loading this library in a service worker and it fails to run because it attempts to use `document` even if it's getCssSelector isn't called.

I think the library should not expect these globals before it's called. 

The same happens if I try to run `build/index.js` in node (just as a test), except it complains about `self`

Can changes be made so the module doesn't break the rest of the app by just "existing"?

**Comment by fregante** [comment]
I think this is the culprit:

https://github.com/fczbkk/css-selector-generator/blob/37466d5b3798f0a6981e760babeced69e3a45e4a/src/utilities-options.ts#L22

**Comment by fczbkk** [comment]
Thanks for the bug report @fregante. I have created a pre-patch version `v3.4.5-0`, that should potentially solve the problem. It removes all references to the `document` object:
https://www.npmjs.com/package/css-selector-generator/v/3.4.5-0

Can you please check if it works for you, before I create official patch build of the library?

**Comment by fregante** [comment]
Thanks! It appears to have fixed it in service workers, but:

### Node still can't run the bundle

… because (I think) webpack is injecting the keyword `self` in it. You can find the keyword here but not on the repo: https://www.unpkg.com/css-selector-generator@3.4.5-0/build/index.js

Maybe updating webpack or reconfiguring it will fix it

### TypeScript is throwing errors 

...with 3.4.5-0, but maybe it was just an incomplete publish:

```

src/nativeEditor/infer.ts:462:3 - error TS2322: Type '"id"' is not assignable to type 'CssSelectorType'.

462   "id",
      ~~~~

src/nativeEditor/infer.ts:463:3 - error TS2322: Type '"tag"' is not assignable to type 'CssSelectorType'.

463   "tag",
      ~~~~~

src/nativeEditor/infer.ts:464:3 - error TS2322: Type '"class"' is not assignable to type 'CssSelectorType'.

464   "class",
      ~~~~~~~

src/nativeEditor/infer.ts:465:3 - error TS2322: Type '"attribute"' is not assignable to type 'CssSelectorType'.

465   "attribute",
      ~~~~~~~~~~~

src/nativeEditor/infer.ts:466:3 - error TS2322: Type '"nthoftype"' is not assignable to type 'CssSelectorType'.

466   "nthoftype",
      ~~~~~~~~~~~

src/nativeEditor/infer.ts:467:3 - error TS2322: Type '"nthchild"' is not assignable to type 'CssSelectorType'.

467   "nthchild",
      ~~~~~~~~~~
```

**Comment by fczbkk** [comment]
Hey @fregante, can you please try `v3.4.6-0`?

I have reorganised the project a bit and added an ESM version. The UMD code generated by Webpack contains references to `self`. I don't want to convert the original UMD build, because that would be a major breaking change. 

You should be able to use the ESM version like this:

```javascript
import getCssSelector from 'css-selector-generator/esm';
```

**Comment by fregante** [comment]
Watch out, `exports` itself is a breaking change because it allows **only** what's listed in it.

The regular old export is now broken. This is the webpack error:

```
Module not found: Error: Package path . is not exported from package ./node_modules/css-selector-generator (see exports field in ./node_modules/css-selector-generator/package.json)
```

Types are also not found on the `/esm` export:

```
src/blocks/readers/frameworkReader.ts:38:40 - error TS7016: Could not find a declaration file for module 'css-selector-generator/esm'. './node_modules/css-selector-generator/esm/index.js' implicitly has an 'any' type.
  If the 'css-selector-generator' package actually exposes this module, try adding a new declaration (.d.ts) file containing `declare module 'css-selector-generator/esm';`

38   const getCssSelector = (await import("css-selector-generator/esm")).default;
                                          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

If you want to try it locally, either set a simple webpack/typescript build or you can check out this repository https://github.com/pixiebrix/pixiebrix-extension and use `npm run build` (it takes about a minute to build)

---

As far as I'm concerned, the previous version was enough (without `document` references) as long as the types were bundled correctly; That already works in Service Workers. If anyone else complains about `self` that can be addressed separately.

**Comment by fregante** [comment]
> Types are also not found on the `/esm` export:

No .d.ts files:
https://www.unpkg.com/browse/css-selector-generator@3.4.6-0/esm/

**Comment by fczbkk** [comment]
Hey @fregante, I have just published `v3.5.0`. It should be fully backwards compatible, but it also contains ESM build. You can use it by importing it like this:

```javascript
import { getCssSelector } from 'css-selector-generator/esm';
```

Both UMD and ESM version should contain type definitions. I have tested using both versions using Webpack and EsBuild. Everything seemed to be working fine.

I'm closing the issue. Feel free to reopen it if you encounter any problem.
